# Table of Contents
 <p><div class="lev1"><a href="# Step 1. Process data"><span class="toc-item-num">1&nbsp;&nbsp;</span>TODO: Following given notebooks in HW1, create table of contents here :) </a></div>

This Jupyter extension might help. https://github.com/minrk/ipython_extensions

# Homework 03 - Interactive Viz

# Overview 

### Objective : Build a [Choropleth map](https://en.wikipedia.org/wiki/Choropleth_map) which shows intuitively how much grant money goes to each Swiss canton.

##### What data do we have?

The SNSF (Swiss National Science Foundation) has graciously provided the [P<sup>3</sup>](http://p3.snf.ch/) database, which contains data on research projects approved by the SNSF. We will just use the [Grants Data](http://p3.snf.ch/P3Export/P3_GrantExport.csv) (saved in `data/P3_GrantExport`) as this contains the grant amounts going to different universities.

We are also given the file `data/ch-cantons.topojson.json`, which contains the geo-coordinates of each Swiss canton.

##### What needs to be done?

The Grants Data does not contain the Canton of the University/Institution that received the funding. Therefore, we will need to deduce the Canton from the University/Institution. Once this is accomplished, we can sum the grant money that was allocated to each Canton.

##### How will this be done?

We will use the following Python packages to accomplish our objective:
1. `pandas` to import, clean, and wrangle the Grants Data
2. `json` and `pprint` to read the `data/ch-cantons.topojson.json` file
3. `requests` with the [GeoNames API](http://www.geonames.org/export/web-services.html) and `googlemaps` (a [Python wrapper](https://github.com/googlemaps/google-maps-services-python) for Google Maps API Web Services) to determine the Canton of a particular University/Institution
4. `folium` to deal with the geographical data and create the Choropleth map

# Building the Choropleth Map

We start off by importing the Python packages mentioned above.

In [1]:
import requests
import folium
import pandas as pd
import numpy as np
import json
import random
from pprint import pprint

## 1. Importing the Data

We downloaded the [Grants Data](http://p3.snf.ch/P3Export/P3_GrantExport.csv) and placed it in the local `data` folder.

In [2]:
GRANTS_FILE = 'data/P3_GrantExport.csv'
grants_data_orig = pd.read_csv(GRANTS_FILE, sep = ';')
grants_data_orig.head(3)

"Project Number"                                      Project Title  \
0                  1  Schlussband (Bd. VI) der Jacob Burckhardt-Biog...   
1                  4  Batterie de tests à l'usage des enseignants po...   
2                  5  Kritische Erstausgabe der "Evidentiae contra D...   

  Project Title English                              Responsible Applicant  \
0                   NaN                                       Kaegi Werner   
1                   NaN                                 Massarenti Léonard   
2                   NaN  Kommission für das Corpus philosophorum medii ...   

             Funding Instrument Funding Instrument Hierarchy  \
0  Project funding (Div. I-III)              Project funding   
1  Project funding (Div. I-III)              Project funding   
2  Project funding (Div. I-III)              Project funding   

                                         Institution  \
0                                                NaN   
1  Faculté de Psychologie et des Sciences de l'Ed...   
2  Kommission für das Corpus philosophorum medii ...   

                                University  Discipline Number  \
0                     Nicht zuteilbar - NA              10302   
1                Université de Genève - GE              10104   
2  NPO (Biblioth., Museen, Verwalt.) - NPO              10101   

                    Discipline Name  \
0                     Swiss history   
1  Educational science and Pedagogy   
2                        Philosophy   

                           Discipline Name Hierarchy  Start Date    End Date  \
0  Human and Social Sciences;Theology & religious...  01.10.1975  30.09.1976   
1  Human and Social Sciences;Psychology, educatio...  01.10.1975  30.09.1976   
2  Human and Social Sciences;Linguistics and lite...  01.03.1976  28.02.1985   

  Approved Amount Keywords  
0        11619.00      NaN  
1        41022.00      NaN  
2        79732.00      NaN

For our objective, we are interested in the following columns: 

* ** University **: the (possible) university proposing the project. As we see in the third row, a grant can be awarded to an NPO (Non-Profit Organization).
* ** Institution **: the institution proposing the project.
* ** Approved Amount **: amount of approved grants

The Institution/University can be used to deduce the Swiss Canton. Finally, the Amount Approved will be used for our Choropleth map. Let's extract these columns to reduce the size of our DataFrame.

In [3]:
grants_data_red = grants_data_orig[['University', 'Institution', 'Approved Amount']]
grants_data_red.head()

University  \
0                     Nicht zuteilbar - NA   
1                Université de Genève - GE   
2  NPO (Biblioth., Museen, Verwalt.) - NPO   
3                   Universität Basel - BS   
4  NPO (Biblioth., Museen, Verwalt.) - NPO   

                                         Institution Approved Amount  
0                                                NaN        11619.00  
1  Faculté de Psychologie et des Sciences de l'Ed...        41022.00  
2  Kommission für das Corpus philosophorum medii ...        79732.00  
3  Abt. Handschriften und Alte Drucke Bibliothek ...        52627.00  
4                       Schweiz. Thesauruskommission       120042.00

Finally, we will rename the "Approved Amount" column to "Amount" so that it can be easily accessed.

In [4]:
grants_data_red = grants_data_red.rename(columns={'Approved Amount': 'Amount'})

## 2. Cleaning the Data

### Cleaning "Amount"

Let's make sure the "Amount" has the correct variable type, i.e. numeric, and that entries with an invalid "Amount" are dropped.

In [5]:
print(grants_data_red.shape)
grants_clean_amount = grants_data_red
grants_clean_amount.Amount = pd.to_numeric(grants_clean_amount.Amount, errors='coerce')
grants_clean_amount = grants_data_red.dropna(subset=["Amount"])
print(grants_clean_amount.shape)

(63969, 3)
(53059, 3)


We have dropped 10910 entries.

### Cleaning "University" and "Institution"

We only need one of these fields to be valid for each row in order to geolocate where the grant was awarded. So let's drop those rows that have invalid entries for **both** fields. Moreover, we have noticed that the "University" entries have the following structure:

`LONG NAME - SHORT NAME`

Furthermore, one of the entries is `Nicht zuteilbar - NA` which must be SNSF's way of indicating an invalid entry. We will first replace such entries with the standard entry for invalids - `NaN` - before dropping rows.

In [6]:
print(grants_clean_amount.shape)
grants_clean_place = grants_clean_amount.replace(to_replace="Nicht zuteilbar - NA", value=nan)
grants_clean_place.dropna(how='all',subset=["University","Institution"], inplace=True)
print(grants_clean_place.shape)

(53059, 3)
(49823, 3)


In [7]:
# suppressed output for shorter notebook!
#grants_clean_place.University.value_counts(dropna=False)

We have dropped 3236 entries. Almost done! We only one need either "Univerisity" or "Institution" for geolocation. For most entries the "University" should be enough. However, 1437 of the entries are not from Universities but rather Non-Profit Organizations (NPO) and there are 1540 `NaN`s. This count was determined by running:

`grants_clean_place.University.value_counts(dropna=False)`

We will modify the "University" column by setting its value to `NaN` if the entry is "NPO (Biblioth., Museen, Verwalt.) - NPO". For the valid universities, we will decouple the `LONG NAME` and `SHORT NAME` and place them in different columns. These modifications will make it easier later on when mapping the university/institution to a Canton and for checking the criteria that 95% of the universities are mapped to a canton.

In [21]:
def extract_short_name(row):
    # extract the SHORT NAME from University
    uni = str(row["University"])
    if uni == "NPO (Biblioth., Museen, Verwalt.) - NPO" or pd.isnull(uni):
        return np.nan
    split_entry = uni.split()
    return split_entry[-1]

def extract_long_name(entry):
    # extract the LONG NAME and remove the space at the end
    split_entry = entry.rsplit("-",1)
    return split_entry[0].rstrip()

def modify_university(row): 
    # keep long name unless it is an NPO or NaN
    if row["University"] == "NPO (Biblioth., Museen, Verwalt.) - NPO" or pd.isnull(row["University"]):
        return np.nan
    else: 
        return extract_long_name(str(row["University"]))

grants_modified_uni = grants_clean_place.copy()
grants_modified_uni['Short'] = grants_modified_uni.apply(lambda row: extract_short_name(row), axis=1)
grants_modified_uni['University'] = grants_modified_uni.apply(lambda row: modify_university(row), axis=1)

In [22]:
grants_modified_uni.head()

University                                        Institution  \
1    Université de Genève  Faculté de Psychologie et des Sciences de l'Ed...   
2                     NaN  Kommission für das Corpus philosophorum medii ...   
3       Universität Basel  Abt. Handschriften und Alte Drucke Bibliothek ...   
4                     NaN                       Schweiz. Thesauruskommission   
5  Université de Fribourg  Séminaire de politique économique, d'économie ...   

     Amount Short  
1   41022.0    GE  
2   79732.0   NaN  
3   52627.0    BS  
4  120042.0   NaN  
5   53009.0    FR

### Summary

We have gone from 63696 entries to 49823 by dropping invalid "Amount" and "University" or "Institution" entries. We have also re-labelled data (NPO to `NaN` and decoupled the long and short name of the university for our convenience in analysis).

## 3. Reading the Canton Data

Now we obtain the abbrevations for all cantons from the geographical JSON data in `data/ch-cantons.topojson.json`. We did some inspection of the structure of the JSON file (using pprint) and extracted the canton abbrevations by navigating appropriately through the JSON data structure.

In [10]:
CH_GEO_JSON = 'data/ch-cantons.topojson.json'

with open(CH_GEO_JSON, 'r') as f:
    canton_data = json.load(f)

cantons = canton_data['objects']['cantons']['geometries']
canton_ids = []
for canton in cantons:
    canton_ids.append(canton['id'])

print(canton_ids)
#pprint(canton_data)

['ZH', 'BE', 'LU', 'UR', 'SZ', 'OW', 'NW', 'GL', 'ZG', 'FR', 'SO', 'BS', 'BL', 'SH', 'AR', 'AI', 'SG', 'GR', 'AG', 'TG', 'TI', 'VD', 'VS', 'NE', 'GE', 'JU']


## 4. Mapping "Institute" to a Canton

We will use a Python wrapper for geolocating the Universities and Institutions. Let's see how it works for `EPFL` so we know how to parse its result!

In [16]:
import googlemaps
# add your own GOOGLE API KEY HERE!!!
key = % env GOOGLE_MAPS_KEY
gmaps = googlemaps.Client(key=key)
test_epfl = gmaps.geocode('EPFL')
test_epfl

[{'address_components': [{'long_name': 'EPFL',
    'short_name': 'EPFL',
    'types': ['premise']},
   {'long_name': 'Lausanne',
    'short_name': 'Lausanne',
    'types': ['locality', 'political']},
   {'long_name': 'Ouest lausannois',
    'short_name': 'Ouest lausannois',
    'types': ['administrative_area_level_2', 'political']},
   {'long_name': 'Vaud',
    'short_name': 'VD',
    'types': ['administrative_area_level_1', 'political']},
   {'long_name': 'Switzerland',
    'short_name': 'CH',
    'types': ['country', 'political']},
   {'long_name': '1015', 'short_name': '1015', 'types': ['postal_code']}],
  'formatted_address': 'EPFL, 1015 Lausanne, Switzerland',
  'geometry': {'location': {'lat': 46.5189865, 'lng': 6.5676007},
   'location_type': 'ROOFTOP',
   'viewport': {'northeast': {'lat': 46.52033548029151,
     'lng': 6.568949680291502},
    'southwest': {'lat': 46.51763751970851, 'lng': 6.566251719708498}}},
  'place_id': 'ChIJ4zm3ev4wjEcRShTLf2C0UWA',
  'types': ['premise']}

The output is a list of possible locations (in this case the list of length 1). So we can loop through the candidates to find the Canton. At `administrative_area_level_1`, we can extract the short code of the Canton (which will correspond to the Canton ID in the geographical JSON data). We will also need to check the country to make sure the university/institution is in Switzerland!

In [12]:
# check if result is in Switzerland
def in_Switzerland(place_info):
    for info_dict in place_info['address_components']:
        if ('country' in info_dict['types']) and (info_dict['long_name'] == 'Switzerland'):
            return True
    return False

in_Switzerland(test_epfl[0])

True

As we can see EPFL is in Switzerland! If the University/Institution is in Switzerland, we can then parse the output for the Canton at `'administrative_area_level_1`.

In [13]:
CANTON_TYPE_KEY = 'administrative_area_level_1'

# parse the Canton short name
def parse_canton(place_candidates):
    if (place_candidates == None or len(place_candidates) == 0):
        return None
    # go through multiple results in given to this function by passing a list of 'geocode' results
    for place in place_candidates:
        if not in_Switzerland(place):
            continue
        for info_dict in place['address_components']:
            if (CANTON_TYPE_KEY in info_dict['types']) and (info_dict['short_name'] in canton_ids):
                return info_dict['short_name']
    return None    

parse_canton(test_epfl)

'VD'

We have noticed that using only the University Long Name is not robust enough. So we will use a combination of the University Long Name, the Institution, and the University Short Name in order to determine the Canton as such:

* IF it is a University: try LONG NAME
    * IF LONG NAME worked: return Canton ID
    * ELSEIF LONG NAME couldn't find a results: try Institution (if it is not NaN)
    * IF still no result: try SHORT NAME + "CH"
    * return NaN if search didn't return a result
* ELSE: try Institution
    * return Canton ID or NaN if search didn't return a result
    
We will do this with a dictionary though, so that we don't have to make 49823 queries! So we try all unique University names and for those Universities that.

In [25]:
universities = grants_modified_uni.University.dropna().unique()
universities

array(['Université de Genève', 'Universität Basel',
       'Université de Fribourg', 'Universität Zürich',
       'Université de Lausanne', 'Universität Bern',
       'Eidg. Forschungsanstalt für Wald,Schnee,Land',
       'Université de Neuchâtel', 'ETH Zürich',
       'Inst. de Hautes Etudes Internat. et du Dév',
       'Universität St. Gallen', 'Weitere Institute',
       'Firmen/Privatwirtschaft', 'Pädagogische Hochschule Graubünden',
       'EPF Lausanne', 'Pädagogische Hochschule Zürich',
       'Universität Luzern',
       'Schweiz. Institut für Kunstwissenschaft - SIK',
       'SUP della Svizzera italiana', 'HES de Suisse occidentale - HES',
       'Robert Walser-Stiftung Bern', 'Paul Scherrer Institut',
       'Pädagogische Hochschule St. Gallen',
       'Eidg. Anstalt für Wasserversorgung',
       'Eidg. Material und Prüfungsanstalt',
       'Physikal.-Meteorolog. Observatorium Davos',
       'Berner Fachhochschule',
       'Swiss Center for Electronics and Microtech.', 'Weite

In [343]:
def search_canton(row):
    # for universities
    if pd.notnull(row.University):
        # first try university
        canton = parse_canton(gmaps.geocode(row.University))
        if canton is not None:
            return canton
        # try institution if University name didn't work
        elif pd.notnull(row.Institution):
            canton = parse_canton(gmaps.geocode(row.Institution))
        if canton is not None:
            return canton
        # last resort - combine short name and "CH"
        else:
            canton = parse_canton(gmaps.geocode(str(row.Short)+"CH"))
    else:  # for research institutes
        canton = parse_canton(gmaps.geocode(row.Institution))
    return canton

In [350]:
grants_modified_uni.Short.unique()

array(['GE', 'NaN', 'BS', 'FR', 'ZH', 'LA', 'BE', 'WSL', 'NE', 'ETHZ',
       'IHEID', 'SG', 'FINST', 'FP', 'PHGR', 'EPFL', 'nan', 'PHZFH', 'LU',
       'SIK-ISEA', 'SUPSI', 'HES-SO', 'RWS', 'PSI', 'PHSG', 'EAWAG',
       'EMPA', 'PMOD', 'BFH', 'CSEM', 'ASPIT', 'AORI', 'SIAF', 'FIBL',
       'FMI', 'KSPSG', 'AGS', 'EOC', 'IST', 'EHB', 'ZFH', 'USI', 'IKG',
       'FHNW', 'HfH', 'CREALP', 'IDIAP', 'PHBern', 'IUKB', 'SPF', 'HSLU',
       'IRO', 'HEPL', 'FHO', 'ISSKA', 'SIB', 'HEPBEJUNE', 'PHLU', 'SAGW',
       'ISR', 'ASP', 'HEPFR', 'PHSZ', 'PHTG', 'BITG', 'FHKD', 'FORS',
       'PHVS', 'SHLR', 'PHZG', 'IRSOL', 'FUS', 'PHSH', 'PHFHNW', 'STHB',
       'FTL', 'FFHS'], dtype=object)

In [344]:
# def find_canton(row):
#     search_canton
#     # if not searched before
#     if row.University not in uni_canton_dict:
#         search_keys = construct_search_keys(row.University, row.Institution)
#         uni_canton_dict[row.University] = search_canton(search_keys)
#         print(row.University + " " + uni_canton_dict[row.University])
#     return uni_canton_dict[row.University]

grants_canton = grants_modified_uni.copy()
grants_canton['Canton'] = grants_canton.apply(search_canton, axis=1)

KeyboardInterrupt: 

In [165]:
grants_canton.Canton.value_counts(dropna=False)

NaN    396
ZH      28
VD      14
BS      13
GE      11
BE      11
FR       9
SG       8
LU       8
TI       6
TG       5
NE       5
SH       3
GR       2
AG       2
OW       1
ZG       1
SO       1
SZ       1
BL       1
Name: Canton, dtype: int64

To compute grants by canton, we need to find the corresponding Canton for a particular "Institute".

We will first use the **Geonames API** as this was suggested by the assignment but then revert to the results of the **GoogleMaps API** as the latter is able to map more of the "Institutes". 

### Geonames API

In [114]:
uni_canton = pd.DataFrame(columns=['University','Canton'])

In [115]:
def canton_in_CH(result):
    for i in range(len(result)):
        if(result['geonames'][i]['countryName']=='Switzerland'):
            return True
    return False

In [116]:
URL = 'http://api.geonames.org/searchJSON?'

def canton_get(uni):
    paraload = {'q':uni,'username':'shiyuenie'}
    r = requests.get(URL, params=paraload)
    result = r.json()
    return result

In [342]:
canton_get("AGRIDEA, CH")

{'geonames': [], 'totalResultsCount': 0}

In [118]:
for ucount in range(len(institutes)): # search full name
    uni = institutes[0][ucount]
    result = canton_get(uni)
    if (result['geonames'] != []):
        if (canton_in_CH(result)):
            uni_canton.loc[ucount] = [uni,result['geonames'][0]['adminName1']]
        else:
            uni_canton.loc[ucount] = [uni,'Not Found']   
    else:
        uni_canton.loc[ucount] = [uni,'Not Found']
    
uni_canton.head()

KeyError: 0

In [ ]:
#uni_name=uni_name.replace({r'\s+': '&'}, regex=True)

for ucount in range(len(uni_name)):
    if (uni_canton.loc[:,'Canton'][ucount] == 'Not Found'): # search with seperate parts
        uni_full = uni_name['Full Name'][ucount] 
        uni_short = uni_name['Short Name'][ucount] 
        result = canton_get(uni_full) # search part 1
        if (result['geonames'] != []):
            if(canton_in_CH(result)):
                uni_canton.loc[ucount] = [uni,result['geonames'][0]['adminName1']]
            else:
                result = canton_get(uni_short) # search part 2 
                if (result['geonames'] != []):
                    if(canton_in_CH(result)):
                        uni_canton.loc[ucount] = [uni,result['geonames'][0]['adminName1']]
        
uni_canton.head()

In [ ]:
uni_canton[uni_canton['Canton'] != "Not Found"]

for ucount in range(len(uni_test)): # search full name wit &
    uni = uni_test[0][ucount]
    result = canton_get(uni)
    if (result['geonames'] != []):
        if (canton_in_CH(result)):
            uni_canton.loc[ucount] = [uni,result['geonames'][0]['adminName1']]
    
uni_canton.head()

In [ ]:
np.sum(uni_canton.loc[:,'Canton']=='Not Found')

In [ ]:
uni_canton[uni_canton['Canton'] != "Not Found"]

In [ ]:
np.sum(uni_canton.loc[:,'Canton']!='Not Found')

## GoogleMaps API

We use a Python [API wrapper](https://github.com/googlemaps/google-maps-services-python) for GoogleMaps

In [ ]:
# to install the wrapper: pip install -U googlemaps

In [ ]:
import googlemaps

# Key omitted due to security concerns on Github
# API_KEY = 'AIzaSyC2-TxJBHd-X8RnFh3-a6Y9hXri7WpktHE'
key = % env GOOGLE_MAPS_KEY
#########################'

gmaps = googlemaps.Client(key=key)

In [ ]:
# We try a sample search to obtain the structure of reply
test_epfl = gmaps.geocode('EPFL Switzerland')
test_epfl

We notice that canton information is returned in 'administrative_area_level_1' field. long_name gives the full name of the canton, short_name gives the abbrevation, the same as the id used in geometry json file.

In [ ]:
CANTON_TYPE_KEY = 'administrative_area_level_1'
    
def place_in_CH(place_info):
    for info_dict in place_info['address_components']:
        if info_dict['long_name'] == 'Switzerland':
            return True
    return False

def parse_canton(place_candidates):
    if (place_candidates == None or len(place_candidates) == 0):
        return None
    
    for place in place_candidates:
        if not place_in_CH(place):
            continue
        for info_dict in place['address_components']:
            if (CANTON_TYPE_KEY in info_dict['types']) and (info_dict['short_name'] in canton_ids):
                return info_dict['short_name']
    
    return None    

parse_canton(test_epfl)

In [ ]:
def construct_search_keys(university, institution):
    
    def keys_with(x):
        keys = []
        keys.append(x)
        keys.append(x + ', Switzerland')
        keys.append(x + ', CH')
        return keys
    
    uni_full_name, uni_short_name = university.rsplit('-',1)

    keys = []   
    keys += keys_with(university)
    keys += keys_with(uni_full_name)
    
    if institution != None:
        keys += keys_with(str(institution))
    if uni_short_name != None:
        keys += keys_with(uni_short_name)
    return keys 

In [ ]:
def search_canton(keys):
    for key in keys:
        resp = gmaps.geocode(key)
        canton = parse_canton(resp)
        if canton != None:
            return canton
    return None

In [ ]:
# intialize uni_canton_dict from file
# uni_canton_dict maps university to canton
uni_canton_df = pd.read_csv('uni-canton.csv', encoding='utf-8')
uni_canton_df.set_index('University', inplace=True)
uni_canton_dict = uni_canton_df.to_dict()['0']

In [ ]:
def find_canton(row):
    # if not searched before
    if row.University not in uni_canton_dict:
        search_keys = construct_search_keys(row.University, row.Institution)
        uni_canton_dict[row.University] = search_canton(search_keys)
        print(row.University + " " + uni_canton_dict[row.University])
    return uni_canton_dict[row.University]

df = data.copy()
df['Canton'] = df.apply(find_canton, axis=1)

In [ ]:
# uni_canton_df = pd.DataFrame.from_dict(uni_canton_dict, orient="index")
# uni_canton_df.index.name = 'University'
# uni_canton_df.to_csv('uni-canton.csv', encoding='utf-8')

In [ ]:
df = data.copy()
df['Canton'] = df.apply(find_canton, axis=1)

In [ ]:
df.head()

In [ ]:
df.to_csv('data-with-canton.csv', encoding='utf-8')

In [ ]:
found_rate = pd.notnull(df['Canton']).sum() / len(df['Canton'])
found_rate

### Manually handle missing data

Fow now, we simply ignore the missing records.

#  Compute grants by canton

In [ ]:
canton_grants_df = df[['Approved Amount', 'Canton']].groupby(['Canton']).mean()
canton_grants_df.reset_index(inplace=True)

In [ ]:
# Add missing data
for canton_id in canton_ids:
    if canton_id not in canton_grants_df['Canton'].values:
        canton_grants_df = canton_grants_df.append({'Canton':canton_id, 'Approved Amount': 0.0}, ignore_index=True)

canton_grants_df = canton_grants_df.sort_values(by=['Approved Amount'], ascending=False)
canton_grants_df

# Data visualization
##  Draw grants map by canton

In [ ]:
map = folium.Map(location=[46.82244,8.22410], zoom_start=8)
map.choropleth(data=canton_grants_df,
               columns=['Canton', 'Approved Amount'], 
               key_on='feature.id', 
               geo_path=CH_GEO_JSON, 
               topojson='objects.cantons', 
               fill_color='YlOrRd'
               )
map

** TODO **: The map cannot be displayed on Github. To view the map, you have to run this cell locally with Jupyter Notebook. To facilitate code reviewing, we may add an additionaly snapshot of the map here.